internal/rbd/clone.go

```diff
@@ -0,0 +1,236 @@
+}
+
+func (rv *rbdVolume) generateTempClone() *rbdVolume {
+	tempClone := rbdVolume{}
+	tempClone.conn = rv.conn
+	// The temp clone image need to have deep flatten feature
+	f := []string{librbd.FeatureNameLayering, librbd.FeatureNameDeepFlatten}
+	tempClone.imageFeatureSet = librbd.FeatureSetFromNames(f)
+	tempClone.ClusterID = rv.ClusterID
+	tempClone.Monitors = rv.Monitors
+	tempClone.Pool = rv.Pool
+	// The temp cloned image name will be always (rbd image name + "-temp")
+	// this name will be always unique, as cephcsi never creates an image with
+	// this format for new rbd images
+	tempClone.RbdImageName = rv.RbdImageName + "-temp"
+	return &tempClone
+}
+
+func (rv *rbdVolume) createCloneFromImage(ctx context.Context, parentVol *rbdVolume) error {
+	// generate temp cloned volume
+	tempClone := rv.generateTempClone()
+	snap := &rbdSnapshot{
+		RbdSnapName: rv.RbdImageName,
+		Pool:        rv.Pool,
+	}
+
+	var (
+		iErr error
+		fErr error
+		err  error
+	)
+	var efip ErrFlattenInProgress
+	var j = &journal.Connection{}
+
+	j, err = volJournal.Connect(rv.Monitors, rv.conn.Creds)
+	if err != nil {
+		return status.Error(codes.Internal, err.Error())
+	}
+	defer j.Destroy()
+
+	// create snapshot and temporary clone and delete snapshot
+	err = createRBDClone(ctx, parentVol, tempClone, snap, rv.conn.Creds)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil || iErr != nil {
+			if !errors.Is(iErr, &efip) {
+				// cleanup snapshot
+				cErr := cleanUpSnapshot(ctx, parentVol, snap, tempClone, rv.conn.Creds)
+				if cErr != nil {
+					klog.Errorf(util.Log(ctx, "failed to cleanup image %s or snapshot %s: %v"), snap, tempClone, cErr)
+				}
+			}
+		}
+		if err != nil || fErr != nil {
+			cErr := cleanUpSnapshot(ctx, tempClone, snap, rv, rv.conn.Creds)
+			if cErr != nil {
+				klog.Errorf(util.Log(ctx, "failed to cleanup image %s or snapshot %s: %v"), snap, tempClone, cErr)
+			}
+		}
+	}()
+	// flatten clone
+	iErr = tempClone.flattenRbdImage(ctx, rv.conn.Creds, false, rbdHardMaxCloneDepth, rbdSoftMaxCloneDepth)
+	if iErr != nil {
+		return err
+	}
+	// create snap of temp clone from temporary cloned image
+	// create final clone
+	// delete snap of temp clone
+	fErr = createRBDClone(ctx, tempClone, rv, snap, rv.conn.Creds)
+	if fErr != nil {
+		// set iErr error to cleanup temporary snapshot and temporary clone
+		iErr = errors.New("failed to create user requested cloned image")
+		return fErr
+	}
+	err = rv.getImageID()
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to get volume id %s: %v"), rv, err)
+		return err
+	}
+
+	err = j.StoreImageID(ctx, rv.JournalPool, rv.ReservedID, rv.ImageID, rv.conn.Creds)
+	if err != nil {
+		klog.Errorf(util.Log(ctx, "failed to store volume %s: %v"), rv, err)
+		return err
+	}
+	return nil
+}
+
+func (rv *rbdVolume) flattenCloneImage(ctx context.Context) error {
+	tempClone := rv.generateTempClone()
+	// reducing the limit for cloned images to make sure the limit is in range
```

Comment:
Why!? I think I understand, but explain the need please. The comment describes what the code does, but not why it is needed.